rainier-core/src/main/scala/com/stripe/rainier/compute/Fn.scala

```diff
@@ -0,0 +1,94 @@
+package com.stripe.rainier.compute
+
+import scala.collection.mutable.ArrayBuffer
+
+trait Fn[-A, +Y] { self =>
+  protected type X
+
+  protected def wrap(a: A): X
+  protected def extract(a: A, acc: List[Double]): List[Double]
+  protected def create(columns: List[Array[Double]]): (X, List[Array[Double]])
+  protected def xy(x: X): Y
+
+  def apply(a: A): Y = xy(wrap(a))
+
+  def encode(as: Seq[A]): Y = {
```

Comment:
I'm not sure as `encode` is a good descriptive name for what this does.

Reply:
Fair yep. Any suggestions?

Reply:
I was afraid you'd ask that.

One other thing that seems odd is that both `apply` and `encode` return a `Y`. Maybe it makes sense to get rid of `apply` and rename `encode` to `apply`? I assume Seq will be the more common case and it's easy to `fn(Seq(v))` if necessary.

Reply:
We still need `apply` as a public API, in particular if you want to use the `Fn` as part of prediction, but just in general if you want to treat it as a normal function. Whereas `encode` is closer to a private API (and in fact we could maybe make it `private[rainier]`), that has to be used with care and I do not expect end users to make use of it.

Reply:
Ah OK, I got my usage mixed up then. I'm probably less concerned about the naming if it can be `private[rainier]`. I think I've lost track of what `encode` is doing here.

Relatedly, do we need `wrap` at all or can we `def apply(a: A): Y = encode(Seq(a))`?

Reply:
@andrew-stripe an example will help here.

Let's take the simple case of `f = Fn.numeric[Double].map{x => x + 1}`.

The result of `f.wrap(1.0)` is `Constant(1.0)`.

The result of `f.apply(1.0)` will be `Constant(2.0)`.

The result of `f.encode(List(1.0))` is a Real that is equivalent to `Placeholder(List(1.0)) + 1`.

Does that help?

Reply:
Ah yea, I was getting tripped up on the fact that the return type is the same in both cases. In the example you described it's always `Real` but those represent different compute graphs between `wrap` and `encode`.